### PR TITLE
nixos/flood: init

### PR DESCRIPTION
**Pull request by thiagokokada**
## Description of changes

Add service module for [`flood`](https://flood.js.org/).

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by thiagokokada** [comment]
@ofborg test flood

**Review by thiagokokada — COMMENTED** [review]
CC @winterqt since you maintain the `flood` package.

File: nixos/modules/services/torrent/flood.nix [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+let
+  cfg = config.services.flood;
+in
+{
+  meta.maintainers = with lib.maintainers; [ thiagokokada ];
+
+  options.services.flood = {
+    enable = lib.mkEnableOption "flood";
+    package = lib.mkPackageOption pkgs "flood" { };
+    openFirewall = lib.mkEnableOption "" // {
```
Comment (AndersonTorres) [inline nixos/modules/services/torrent/flood.nix:12]: What does this line do?

Comment (thiagokokada) [reply nixos/modules/services/torrent/flood.nix:12]: It opens the firewall ports: https://github.com/NixOS/nixpkgs/pull/317530/files#diff-9cac63135951b34a3d402ec244c73ab992337c7eec80645de5a755b982c25371R110-R112

**Comment by thiagokokada** [comment]
@ofborg test flood

File: nixos/modules/services/torrent/flood.nix [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    systemd.services.flood = {
+      description = "A web UI for rTorrent with a Node.js backend and React frontend.";
+      after = [ "network.target" ];
+      wantedBy = [ "multi-user.target" ];
+      unitConfig = {
+        Documentation = "https://github.com/jesec/flood/wiki";
+      };
+      serviceConfig = {
+        User = cfg.user;
+        Group = cfg.group;
```
Comment (mweinelt) [inline nixos/modules/services/torrent/flood.nix:71]:
```suggestion
    systemd.services.flood = {
      description = "A web UI for rTorrent with a Node.js backend and React frontend.";
      after = [ "network.target" ];
      wantedBy = [ "multi-user.target" ];
      unitConfig = {
        Documentation = "https://github.com/jesec/flood/wiki";
      };
      serviceConfig = {
        DynamicUser = true;
```

There should be no reason for a static user/group allocation.

File: nixos/modules/services/torrent/flood.nix [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    user = lib.mkOption {
+      type = lib.types.str;
+      description = "User account used to run flood.";
+      default = "flood";
+    };
+    group = lib.mkOption {
+      type = lib.types.str;
+      description = "Group account used to run flood.";
+      default = "flood";
+    };
```
Comment (mweinelt) [inline nixos/modules/services/torrent/flood.nix:25]:
```suggestion
```

File: nixos/modules/services/torrent/flood.nix [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    port = lib.mkOption {
+      type = lib.types.int;
+      description = "Port to bind webserver.";
+      default = 3000;
+    };
+    host = lib.mkOption {
+      type = lib.types.str;
+      description = "Host to bind webserver.";
+      default = "0.0.0.0";
```
Comment (mweinelt) [inline nixos/modules/services/torrent/flood.nix:34]: - Can this bind dualtstacked? For example using `::`
- Can we default to localhost, just so people don't accidentally expose this?

File: nixos/modules/services/torrent/flood.nix [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    };
+    runDir = lib.mkOption {
+      type = lib.types.path;
+      description = "Directory to store Flood configuration. Needs to be readable/writable by Flood's user/group.";
+      default = "/var/lib/flood";
+    };
```
Comment (mweinelt) [inline nixos/modules/services/torrent/flood.nix:40]:
```suggestion
```

Use `StateDirectory = "flood";` instead in the systemd unit. There should be no relevant state here.

File: nixos/modules/services/torrent/flood.nix [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+        RestartSec = "3s";
+        ExecStart = lib.concatStringsSep " " ([
+          "${lib.getExe cfg.package}"
+          "--host"
+          cfg.host
+          "--port"
+          (toString cfg.port)
+          "--rundir"
+          cfg.runDir
+        ] ++ cfg.extraArgs);
```
Comment (mweinelt) [inline nixos/modules/services/torrent/flood.nix:84]:
```suggestion
        ExecStart = utils.escapeSystemdExecArgs ([
          (lib.getExe cfg.package)
          "--host"
          cfg.host
          "--port"
          (toString cfg.port)
          "--rundir"
          cfg.runDir
        ] ++ cfg.extraArgs);
```

Make sure we properly escape the exec args.

**Review by mweinelt — COMMENTED** [review]
I think we can make this service much more lean with a dynamic user and statedir allocation, since it does not actually have relevant state that goes beyond its configuration.

File: nixos/modules/services/torrent/flood.nix (hunk 2 above)
Comment (mweinelt) [reply nixos/modules/services/torrent/flood.nix:71]: And actually the project description is

>  A modern web UI for various torrent clients with a Node.js backend and React frontend.

File: nixos/modules/services/torrent/flood.nix (hunk 4 above)
Comment (thiagokokada) [reply nixos/modules/services/torrent/flood.nix:34]: - Not sure.
- Yes, good idea.

File: nixos/modules/services/torrent/flood.nix (hunk 2 above)
Comment (thiagokokada) [reply nixos/modules/services/torrent/flood.nix:71]: > There should be no reason for a static user/group allocation.

One issue is that in some cases you want a static user/group allocation, for example to reuse a `rtorrent` user to access its permission. In this case you can't use `DynamicUser`.

So I would prefer to keep the service as currently is.

Comment (mweinelt) [reply nixos/modules/services/torrent/flood.nix]: The proper solution for this is `SupplementaryGroup = [ "rtorrent" ]` on the unit IMO.

Comment (thiagokokada) [reply nixos/modules/services/torrent/flood.nix]: @mweinelt Did all the proposed changes. I think in the worst case I can override the service for my own needs. Can you re-review?

File: nixos/modules/services/torrent/flood.nix (hunk 5 above)
Comment (thiagokokada) [reply nixos/modules/services/torrent/flood.nix]: Done.

File: nixos/modules/services/torrent/flood.nix (hunk 3 above)
Comment (thiagokokada) [reply nixos/modules/services/torrent/flood.nix]: Done.

File: nixos/modules/services/torrent/flood.nix [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+      unitConfig = {
+        Documentation = "https://github.com/jesec/flood/wiki";
+      };
+      serviceConfig = {
+        Type = "simple";
+        KillMode = "process";
+        Restart = "on-failure";
+        RestartSec = "3s";
+        ExecStart = utils.escapeSystemdExecArgs ([
+          "${lib.getExe cfg.package}"
```
Comment (mweinelt) [inline nixos/modules/services/torrent/flood.nix:48]:
```suggestion
          (lib.getExe cfg.package)
```

File: nixos/modules/services/torrent/flood.nix [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+  config = lib.mkIf cfg.enable {
+    systemd.services.flood = {
+      description = "A modern web UI for various torrent clients.";
+      after = [ "network.target" ];
+      wantedBy = [ "multi-user.target" ];
+      unitConfig = {
+        Documentation = "https://github.com/jesec/flood/wiki";
+      };
+      serviceConfig = {
+        Type = "simple";
```
Comment (mweinelt) [inline nixos/modules/services/torrent/flood.nix:43]:
```suggestion
```

This is the default.

File: nixos/modules/services/torrent/flood.nix [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+        KillMode = "process";
```
Comment (mweinelt) [inline nixos/modules/services/torrent/flood.nix:44]: This is not recommended. Do you have a specific reason for this setting?

> Note that it is not recommended to set KillMode= to process or even none, as this allows processes to escape the service manager's lifecycle and resource management, and to remain running even while their service is considered stopped and is assumed to not consume any resources.

https://www.freedesktop.org/software/systemd/man/latest/systemd.kill.html#KillMode=

File: nixos/modules/services/torrent/flood.nix [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+        ExecStart = utils.escapeSystemdExecArgs ([
+          "${lib.getExe cfg.package}"
+          "--host"
+          cfg.host
+          "--port"
+          (toString cfg.port)
+          "--rundir=/var/lib/flood"
+        ] ++ cfg.extraArgs);
+
+        CapabilityBoundingSet = "";
```
Comment (mweinelt) [inline nixos/modules/services/torrent/flood.nix:56]:
```suggestion
        CapabilityBoundingSet = [ "" ];
```

This is better, since lists merge, but the default remains an explicitly empty set.

File: nixos/modules/services/torrent/flood.nix [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+        ProtectKernelTunables = true;
+        ProtectProc = "invisible";
+        ProtectSystem = "strict";
+        RestrictAddressFamilies = [ "AF_UNIX" "AF_INET" "AF_INET6" ];
+        RestrictNamespaces = true;
+        RestrictRealtime = true;
+        RestrictSUIDSGID = true;
+        StateDirectory = "flood";
+        SystemCallArchitectures = "native";
+        SystemCallFilter = [ "~@privileged" ];
```
Comment (mweinelt) [inline nixos/modules/services/torrent/flood.nix:77]:
```suggestion
        SystemCallFilter = [ "@system-service" "~@privileged" ];
```

Set an explicity allowed scope first, then deny everything you want to exclude. Else it is `everything, except @privileged`, which is a much broader scope.

File: nixos/modules/services/torrent/flood.nix (hunk 9 above)
Comment (thiagokokada) [reply nixos/modules/services/torrent/flood.nix:44]: Recommendation from upstream: https://github.com/jesec/flood/wiki/Install-Flood-as-a-service-(systemd).

But maybe this is just something random and it doesn't actually need it. I will remove it.

File: nixos/modules/services/torrent/flood.nix (hunk 11 above)
Comment (thiagokokada) [reply nixos/modules/services/torrent/flood.nix:77]: I am almost sure that I did this on purpose, because I was playing with hardening for this service and what I got above was pretty much the most I could harden without breaking it.

But let me try again.

Comment (thiagokokada) [reply nixos/modules/services/torrent/flood.nix:75]: @mweinelt Adding `@system-service` causes the service to instant crash.

Comment (mweinelt) [reply nixos/modules/services/torrent/flood.nix:75]: That is because the scope is now narrower. Check dmesg why it was killed.

Comment (mweinelt) [reply nixos/modules/services/torrent/flood.nix:75]: Suggestion updated.

Comment (thiagokokada) [reply nixos/modules/services/torrent/flood.nix]: Done.

**Comment by thiagokokada** [comment]
@ofborg test flood

**Comment by mweinelt** [comment]
I wish flood worked for Deluge :weary:

**Review by AndersonTorres — APPROVED** [review]

File: nixos/modules/services/torrent/flood.nix [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+{
+  meta.maintainers = with lib.maintainers; [ thiagokokada ];
+
+  options.services.flood = {
+    enable = lib.mkEnableOption "flood";
+    package = lib.mkPackageOption pkgs "flood" { };
+    openFirewall = lib.mkEnableOption "" // {
+      description = "Whether to automatically open the necessary ports in the firewall.";
+      default = true;
+    };
```
Comment (mweinelt) [inline nixos/modules/services/torrent/flood.nix:15]: It is convention that `openFirewall` options default to false except in the case of sshd.

Comment (thiagokokada) [reply nixos/modules/services/torrent/flood.nix]: Done, also added examples for `port` and `host` and updated the description in `openFirewall` to be more informative.

**Comment by thiagokokada** [comment]
@ofborg test flood

**Comment by thiagokokada** [comment]
@mweinelt Are there any other concerns in this PR? Can I get an approval here?

**Review by mweinelt — APPROVED** [review]
